Use std::span more in CSS code

cdumez opened the pull request:
#### fa8f32d70a44bf70b4638aa6b169e075b8ceef53
<pre>
Use std::span more in CSS code
<a href="https://bugs.webkit.org/show_bug.cgi?id=272956">https://bugs.webkit.org/show_bug.cgi?id=272956</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::equalSpans):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::hasPrefix):
(WebCore::isAppleLegacyCSSValueKeyword):
(WebCore::cssValueKeywordID):
* Source/WebCore/css/process-css-pseudo-selectors.py:
* Source/WebCore/css/process-css-values.py:
(GenerationContext):

Canonical link: <a href="https://commits.webkit.org/277766@main">https://commits.webkit.org/277766@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/81a4f294f64a72061c3b4f32b982e53fcc17dc45

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48488 "Passed style check") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27700 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51448 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51176 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44553 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50794 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33636 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25223 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/51176 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49070 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/33636 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41852 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/51176 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/33636 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43025 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6545 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41785 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-jsc") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/33636 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43508 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53081 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47979 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23535 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/51/builds/25223 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/53081 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24800 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42051 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/53081 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25605 "Built successfully") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/55474 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-jsc") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6912 "Built successfully and passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24523 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/55474 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-jsc") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -691,9 +691,10 @@ std::span<uint8_t, Extent == std::dynamic_extent ? std::dynamic_extent: Extent *
+template<typename T, std::size_t TExtent, typename U, std::size_t UExtent>
+bool equalSpans(std::span<T, TExtent> a, std::span<U, UExtent> b)
 {
+    static_assert(sizeof(T) == sizeof(U));
```
We should probably be static asserting or using concepts to make sure we get a compile error if you try to call this when `T` or `U` have non-trivial `==` implementations?

cdumez replied:
I am not sure how to check for non-trivial `operator==` implementations. Maybe I could static_assert for is_standard_layout?
https://en.cppreference.com/w/cpp/types/is_standard_layout

cdumez replied:
The documentation for `std::memcmp` says it is typically meaningful for only trivially-copyable objects that have no padding. So I could check for `std::is_trivially_copy_constructible`.

darinadler replied:
If there is any padding then `memcmp` won’t work. We do this kind of check in `HashTraits`, not by checking what is illegal, but by only allowing it on certain things that are known to be legal.

Wording on `cppreference.com` documentation for `memcmp` calls this “trivially-copyable objects that have no padding”, and while that’s a pretty indirect way to define it, I think that’s what we are trying to check. I’ll try to research this a tiny bit more.

darinadler replied:
`std:: is_trivially_copyable_v` does exist, but not sure how to tackle the padding topic

cdumez replied:
https://en.cppreference.com/w/cpp/types/has_unique_object_representations seems potentially interesting too.

darinadler replied:
Maybe for now we can do a too-conservative padding test by just checking `std::is_scalar_v`?

darinadler replied:
Oh, `has_unique_object_representations` is probably great. Honestly, we don’t have to get this 100% right, I just don’t want someone to write future bad code by accident by using this on something like `WTF::String`.

darinadler commented on Source/WebCore/css/process-css-pseudo-selectors.py:
```diff
@@ -435,37 +435,37 @@ def prefix_value(prefix, value):
+        static inline const SelectorPseudoClassOrCompatibilityPseudoElementEntry* findPseudoClassAndCompatibilityElementName(std::span<const LChar> characters)
         {
-            return SelectorPseudoClassAndCompatibilityElementMapHash::in_word_set(reinterpret_cast<const char*>(characters), length);
+            return SelectorPseudoClassAndCompatibilityElementMapHash::in_word_set(reinterpret_cast<const char*>(characters.data()), characters.size());
```
Surprised you didn’t change this to take a `span` too, given it’s probably not widely used, so wouldn’t balloon the size of the pull request.

cdumez replied:
I had trouble finding the definition of in_word_set, seems to be a gperf function. I'll try and find it again to see if I can modify its declaration.

cdumez replied:
I don't think I can control the parameters of `in_word_set`. gperf seems to allow tweaking the name of the lookup function but not its parameters. If you know how to do it, please let me know.

darinadler replied:
I see, I didn’t realize it was generated by `gperf`. Let’s not waste more time tweaking this.

darinadler commented on Source/WebCore/css/process-css-pseudo-selectors.py:
```diff
@@ -435,37 +435,37 @@ def prefix_value(prefix, value):
                 buffer[i] = static_cast<LChar>(character);
             }}
-            return findPseudoClassAndCompatibilityElementName(buffer, length);
+            return findPseudoClassAndCompatibilityElementName(std::span {{ buffer }}.first(characters.size()));
```
It’s amazing how tiny style things can catch my eye. I like `{{ }}`, maybe even prefer it, but we typically do `{ { } }` instead.

cdumez replied:
Not sure `{ {` would work here. The double curly bracket is purely due to python escaping.

cdumez replied:
I could use `std::span(buffer)` to avoid python escaping if you'd prefer.

darinadler replied:
It’s fine. I don’t know what python escaping is.

cdumez replied:
`{foo}` is used for formatting strings in python, kinda like "%s". To clarify that we want to use '{' in the string instead of formatting, we need to double the curly brackets: `{{`.

https://stackoverflow.com/questions/5466451/how-do-i-escape-curly-brace-characters-in-a-string-while-using-format-or

darinadler replied:
Oh, I didn’t realize this was in a python script. I thought it was a `.cpp` file. Wow, I missed a lot!

cdumez committed:
Use std::span more in CSS code
https://bugs.webkit.org/show_bug.cgi?id=272956

Reviewed by Darin Adler.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::equalSpans):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::hasPrefix):
(WebCore::isAppleLegacyCSSValueKeyword):
(WebCore::cssValueKeywordID):
* Source/WebCore/css/process-css-pseudo-selectors.py:
* Source/WebCore/css/process-css-values.py:
(GenerationContext):

Canonical link: https://commits.webkit.org/277766@main
